### PR TITLE
Concrete method types, import cleanup tweaks, fixes...Oh my

### PR DESCRIPTION
**Pull request by bfcarpio**
A rather large commit that's supposed to be in tandem with @hhursev's default value error handling as mentioned in #318.

I made this PR with the following steps:
1. Used `sed` to add type hints to the concrete scraper classes for the more common methods.
2. `autoimport`ed the types from the standard library
3. `isort`ed the imports because `autoimport` can be overly verbose
4. Updated tests and ensured `mypy` is mostly happy
5. Made small fixes along the way with line endings.

I started making some headway into what's discussed in #283. Some concrete scrapers had `instructions()` return a list so I moved the implementation into `_instructions_list()` and made `instructions()` wrap it. If we ever move to a full implementation I can just whip out `sed` again.

I did a quick run through of searching for any time we returned `""`, `0` or `-1` and adjusted things to return `None`. We may still find more of these in the future.

TODO:

- [x] Check scrapers to ensure `""` and `-1` etc aren't being returned.
- [x] Write friendly comments in the templates so that people know to use `None`.

EDIT:
Saving the formatting and typing script here just in case:
```bash
#!/bin/bash

add_types () {
	sed \
		-i \
		-e 's/host(self):/host(self)->str:/g' \
		-e 's/title(self):/title(self)->Optional[str]:/g' \
		-e 's/total_time(self):/total_time(self)->Optional[int]:/g' \
		-e 's/yields(self):/yields(self)->Optional[str]:/g' \
		-e 's/image(self):/image(self)->Optional[str]:/g' \
		-e 's/nutrients(self):/nutrients(self)->Optional[Dict[str,Any]]:/g' \
		-e 's/language(self):/language(self)->Optional[str]:/g' \
		-e 's/instructions(self):/instructions(self)->Optional[str]:/g' \
		-e 's/ingredients(self):/ingredients(self)->Optional[List[str]]:/g' \
		-e 's/author(self):/author(self)->Optional[str]:/g' \
		-e 's/ratings(self):/ratings(self)->Optional[float]:/g' \
		$1
}

ls recipe_scrapers/*.py | while read file; do add_types "$file"; done

autoimport recipe_scrapers/*.py
isort recipe_scrapers/*.py
black recipe_scrapers/*.py
```

### CODE REVIEW

**Comment by jayaddison** [comment]
Fantastic!  Excited to see this @bfcarpio.  Am starting to take a look now; it could be tomorrow before I have time to perform a 'full' review.

**Comment by jayaddison** [comment]
(ah - I didn't spot that this is in a draft state; either way, I'll plan to take a look out of curiosity)

**Comment by bfcarpio** [comment]
I always appreciate your feedback so no worries, @jayaddison.

I'm not sure how #329 will effect these changes so we might have to play around with some rebasing when the time comes.

File: recipe_scrapers/matprat.py [hunk 1]
```diff
@@ -1,28 +1,24 @@
...
         if tt:
-            tt1 = normalize_string(tt.get_text())
-            tt2 = get_minutes(tt1)
-            if tt1 and (tt2 == 0):
-                total_time = tt1
-            else:
-                total_time = tt2
-        return total_time
+            tt = get_minutes(normalize_string(tt.get_text()))
+        return tt
```
Comment (jayaddison) [inline recipe_scrapers/matprat.py:19]: Looks like a good little cleanup :+1:

File: recipe_scrapers/_utils.py [hunk 2]
```diff
@@ -93,10 +93,11 @@ def url_path_to_dict(path):
...
+    url_dict = matches.groupdict() if matches else None
 
     return url_dict
 
 
-def get_host_name(url):
-    return url_path_to_dict(url.replace("://www.", "://"))["host"]
+def get_host_name(url: str) -> str:
+    data = url_path_to_dict(url.replace("://www.", "://"))
+    return data.get("host", "UNKNOWN HOST") if data else "UNKNOWN HOST"
```
Comment (jayaddison) [inline recipe_scrapers/_utils.py:103]: What's the goal of the unknown host fallback case here?

**Review by jayaddison — COMMENTED** [review]
Progressing well so far - a couple of thoughts:

- Review is a little tricky with formatting and logic changes in the same pull request; it's possible to look into each of the individual commits, but perhaps it'd make sense to apply search-and-replace style changes (anything that shouldn't affect the meaning of the code) in one pull request, and logical changes in another?

- Although consistency's useful, it might not make so much sense to update line-endings and whitespace in the context of test data; if we update the content and tests to be consistent there, then we've created tests that don't correspond to the original web content.

The hostname-related change is worth a bit of attention too.

File: templates/scraper.py [hunk 3]
```diff
@@ -1,31 +1,35 @@
+from typing import List, Optional
+
 from ._abstract import AbstractScraper
 
 
 class Template(AbstractScraper):
+    # Functions should return `None` if they cannot access
+    # the correct value as oppposed to `""` or `-1` etc.
```
[the remark this answers is not in the document]
Comment (bfcarpio) [reply templates/scraper.py:8]: I'm happy to express this information in any way, I thought this would be the most visible to people who just want to create a scraper as fast as possible.

File: recipe_scrapers/_utils.py (hunk 2 above)
Comment (bfcarpio) [reply recipe_scrapers/_utils.py:103]: `url_path_to_dict` has a chance to return `None` due to the regex function. It's probably unlikely for this to be a problem since this is part of the generator code, but I wanted to handle it. I can remove it if need be.

File: recipe_scrapers/matprat.py (hunk 1 above)
Comment (bfcarpio) [reply recipe_scrapers/matprat.py:19]: I think this pattern exists for a couple of the scrapers we have. If we can get that scraper health dashboard up and going I think we'll be able to quickly catch complex code like this.

It turns out that the schema code is very `mypy` friendly, but other scrapers are less likely to be.

**Comment by bfcarpio** [comment]
I completely acknowledge that this PR is huge. I suppose that since we don't do a `mypy` check in our pipelines that it would be fine to merge in the build-generated code by itself.

>  it might not make so much sense to update line-endings and whitespace

Pardon my lack of knowledge, but how do line endings work on web provided content? Are line endings sent based on the original file being served or it it based on the device doing the serving?

In specific context of this project, I've found correlation between who wrote the scraper and what line endings the files are in. I went on a crusade of line endings in an attempt to standardize the repo, but if there's a chance that changing the HTML will change the output of tests then I can revert the changes.

**Comment by jayaddison** [comment]
> I suppose that since we don't do a `mypy` check in our pipelines that it would be fine to merge in the build-generated code by itself.

Good point, hmm.  What was the level of effort to introduce the `mypy` annotations and related fixups?  (to help gauge whether it's better to continue to move forward with it, or to re-apply it in a follow-up changeset)

> Pardon my lack of knowledge, but how do line endings work on web provided content? Are line endings sent based on the original file being served or it it based on the device doing the serving?

It's probably complicated, and I'm not sure I can provide a complete answer :)

Roughly speaking though, if and when a (versioned) test data archive file requires an update, I think we want to replace it completely with an updated copy that we've retrieved, rather than attempting edits within the existing version.  Perhaps that's debatable if there are standardized format migrations going on, but I'm not sure that's the case for web content and line-endings in particular.

> In specific context of this project, I've found correlation between who wrote the scraper and what line endings the files are in. I went on a crusade of line endings in an attempt to standardize the repo, but if there's a chance that changing the HTML will change the output of tests then I can revert the changes.

Yep, for sure - it'll likely be related to individuals' editor settings, and I think we should allow people to work in the kind of environment they want to, but it'd also be nice if the resulting output can be standardized for ease of comparison, review and maintenance.  Do you know whether those `gitattributes` you mentioned could help out with that?  (I'm yet to read up on that in detail.. hopefully soon!)

**Comment by bfcarpio** [comment]
> What was the level of effort to introduce the mypy annotations and related fixups

Now that I know what needs to be done, I can write a quick bash script in 5- 10 minutes and I can do all the automated formatting stuff. Going in and correcting the `mypy` errors (and affected tests) might take another 30-45 minutes. So about an hour total, but it's monotonous for the most part.

> I think we want to replace it completely with an updated copy that we've retrieved, rather than attempting edits within the existing version. 

When you phrase it like that, I whole-hardheartedly agree. I'll do the minimum it takes to my merges in and then we an do the rest as individual PRs.

> gitattributes

Yes! This is where they apply. I'm used to being the lone Linux dev with a bunch of Windows users so I have experience with maintaining repos. `gitattributes` allows the repo to use `lf` (as required etc) and automatically converts to and from your OS's preferred ending when you push and pull. There are other features such as ensuring binaries get treated as binaries and you don't have massive diffs. I consider the a basic `gitattributes` for line endings to be mandatory because you never know when a rogue Windows user will come along. :joy: 

`editorconfigs` have compatible editors (plugins available for most popular ones) change their character, indentation and other settings for each project. I find this especially useful for Python and Yaml due to the indentation requirements, but formatting tools like `black` are even better. Though I did have an issue where a user was using ASCII instead of utf-8 and an `editorconfig` was the simple fix. Not sure what software has ASCII as default these days...

TLDR: 2 small text files can save a lot of headaches and time so might as well, IMHO.

**Comment by bfcarpio** [comment]
@jayaddison I just force pushed a reorganization of commits. If you look at the files view there's a way to see changes in each commit in the top left.

All the automated changes are in the first commit. Then each commit is an incremental change of a certain type. Fixing `instructions()` that returned `List` then adjust defaults to `None` etc.

I've removed the line ending changes and haven't touched any `html` files.

I also realized that some of our tests silenced errors to get the defaults. I've made it so that they throw the exceptions and appropriately ensured that they throw. This is important to note because `_schemaorg.ratings()` has an exception built into it, but the others don't... potential bug worth another PR?

File: recipe_scrapers/_schemaorg.py [hunk 4]
```diff
@@ -46,27 +50,31 @@ def __init__(self, page_data):
...
-            cook_time = get_minutes(self.data.get("cookTime")) or 0
-            total_time = prep_time + cook_time
+        if total_time is None:
+            times = [get_minutes(self.data.get(p)) for p in ["prepTime", "cookTime"]]
+
+            total_time = (
+                None
+                if all([b is None for b in times])
+                else sum([t for t in times if t is not None])
+            )
```
Comment (jayaddison) [inline recipe_scrapers/_schemaorg.py:74]: Idea: could potentially be simplified to:

```py
sum([t for t in times if t]) or None
```

File: tests/test_southernliving.py [hunk 5]
```diff
@@ -52,14 +51,16 @@ def test_ingredients(self):
 
     def test_instructions(self):
         return self.assertEqual(
-            [
```
Comment (jayaddison) [inline tests/test_southernliving.py:55]: Interesting!  We already had one scraper that was returning instructions as a list? :)

File: recipe_scrapers/greatbritishchefs.py [hunk 6]
```diff
@@ -28,15 +31,15 @@ def total_time(self):
...
-    def yields(self):
+    def yields(self) -> Optional[str]:
         recipe_yield = self.soup.find("span", {"class": "RecipeAttributes__Serves"})
         if recipe_yield:
-            yields = normalize_string(
+            recipe_yield = normalize_string(
                 recipe_yield.find("span", {"class": "header-attribute-text"}).get_text()
             )
-            return yields
+        return recipe_yield
```
Comment (jayaddison) [inline recipe_scrapers/greatbritishchefs.py:40]: Taking a mental note to look at some of this logic again during #329.

File: recipe_scrapers/justbento.py [hunk 7]
```diff
@@ -7,46 +9,47 @@ class JustBento(AbstractScraper):
...
-                instructions.append(element)
+                instructions.append(element.get_text())
+            instructions = [
+                normalize_string(instruction) for instruction in instructions
+            ]
 
-        return "\n".join(
-            [normalize_string(instruction.get_text()) for instruction in instructions]
-        )
+        return "\n".join(instructions) if instructions else None
```
Comment (jayaddison) [inline recipe_scrapers/justbento.py:50]: I'm not exactly sure how, yet, but I think/hope we'll be able to find a more Pythonic way to avoid repeated 'if x else None` statements like this.  In some cases `or None` might help, but overall I think there should be a way for us to reach `return x` (single statement without trailing conditionals) eventually.  We don't have to do that in this pull request or the near future, it's just something to think about.

File: recipe_scrapers/_abstract.py [hunk 8]
```diff
@@ -1,15 +1,14 @@
 from collections import OrderedDict
 from json.decoder import JSONDecodeError
+from typing import Any, Dict, List, Optional, Tuple, Union
```
Comment (jayaddison) [inline recipe_scrapers/_abstract.py:3]: If it's not a massive pain, what do you think about splitting these multi-reference imports into multi-line import format?

For example:

```py
from typing import (
    Any,
    Dict,
    List,
    Optional,
    Tuple,
    Union,
)
```

That's based on some ideas I picked up from https://github.com/asottile/reorder_python_imports about reducing conflicts.

(personally I prefer the look of references-within-parens to reduce repetition of the `typing` module reference, but I understand why having standalone import lines makes sense, too)

**Review by jayaddison — APPROVED** [review]
This looks good to me, thanks @bfcarpio!  Some ideas/comments for your consideration - no changes required, though.

File: recipe_scrapers/_abstract.py (hunk 8 above)
Comment (bfcarpio) [reply recipe_scrapers/_abstract.py:3]: I'm going to politely say that this is topic for another PR and issue. I have no preference here, but I made #332 just to force consistency and catch silly errors. Feel free to change it how you see fit.

File: recipe_scrapers/justbento.py (hunk 7 above)
Comment (bfcarpio) [reply recipe_scrapers/justbento.py:50]: I'll admit that I'm unfamiliar with all the ins and outs of using `or` like that. I'll take a look. Some of the ternary statements are important since `0.0` and `0` are `falsey`.

```python
>>> "" or None
>>> 0 or None
>>> None or 0
0
>>> None or ""
''
>>> [] or None
>>> None or []
[]
```

File: recipe_scrapers/greatbritishchefs.py (hunk 6 above)
Comment (bfcarpio) [reply recipe_scrapers/greatbritishchefs.py:40]: Please do! We might need to adjust more of the schema logic for the defaulting.

File: tests/test_southernliving.py (hunk 5 above)
Comment (bfcarpio) [reply tests/test_southernliving.py:55]: Yup! There were at least 4.

File: recipe_scrapers/_schemaorg.py (hunk 4 above)
Comment (bfcarpio) [reply recipe_scrapers/_schemaorg.py:74]: I went to bed thinking about that code and I think it caused my nightmares.

I'm going to redo it later today, but we have to be careful about values being `falsey`. I want to return zero if the website has zeros in it, but if the website has `null` or is empty then we can return `None`. I'll add tests for this.

```python
>>> times = [None, None] # should return None
>>> sum([t for t in times if t]) or None
>>> times = [0, None] # should return 0
>>> sum([t for t in times if t]) or None
>>> times = [0, 0] # should return 0
>>> sum([t for t in times if t]) or None
>>> sum([t for t in times if t is not None]) or None
>>> sum([t for t in times if t is not None])
0 # desired result since website "had" zeroes.
>>> sum([])
0 # bad. I need to check if a filtered times is empty
```

File: recipe_scrapers/justbento.py (hunk 7 above)
Comment (jayaddison) [reply recipe_scrapers/justbento.py:50]: There's probably a more formal description, but it's basically: is the left-hand-side truthy?  If so, return that.  Otherwise, return the right hand side.

File: recipe_scrapers/_abstract.py (hunk 8 above)
Comment (jayaddison) [reply recipe_scrapers/_abstract.py:3]: Cool, yep - makes sense.  I might have a look into opening a PR for this soon.

**Comment by bfcarpio** [comment]
@hhursev I'm going to leave this one to you. I'm not sure if you want to merge it into a dev branch of your own or right into master. Please take a hard look at it and let me know if there's anything I should change.

**Comment by hhursev** [comment]
Apologies for the radio silence on my end last week. Also for not delivering the exception handling updates as promised.

This PR looks good to me and will most definitely be part of the forthcoming major version 13. 🍺
Nothing to add or edit. The code seems to be taken care of quite well.  💅

I'll be merging other misc PRs until v13 is ready (2 weeks from now at most 🤞).

**Comment by bfcarpio** [comment]
No worries, @hhursev. I'll check back in a little bit and do another "absorb master" commit. Say in a week right before you end up merging everything.
